### PR TITLE
Replyto and return path

### PR DESCRIPTION
**Pull request by alex-mo**
* Reply to fix
* set a *valid* Return-Path
* Tests

NB: this changes the behaviour with `Reply-To` (it is not enforced to the variable any more)

### CODE REVIEW

**Comment by alex-mo** [comment]
More changes:
- refactor and test the mail pattern matching
- fix for long TLD
- change behaviour with multiple from addresses
- tests for replyto and multiple from addresses
- Drupal comment style

File: email_check.module [hunk 1]
```diff
@@ -21,47 +23,102 @@
...
     $site_reply_to = variable_get('site_replyto_mail', variable_get('site_mail'));
+    $site_return_path = variable_get('site_mail_return_path', variable_get('site_mail'));
 
     $site_mail_domain  = empty($site_mail_domain) ? 'campaignion.org' : $site_mail_domain;
     $site_mail_address = empty($site_mail_address) ? 'info@campaignion.org' : $site_mail_address;
 
-    // make shure site_mail and site_mail_domain are consistend
+    $reply_to = !empty($message['headers']['Reply-To']) ? $message['headers']['Reply-To'] : '';
+
+    // make sure site_mail and site_mail_domain are consistend
```
Comment (torotil) [inline email_check.module]: Comments should be whole sentences. (phpcs-drupal)

File: email_check.module [hunk 2]
```diff
@@ -21,47 +23,102 @@
...
+    $message_from_addresses = _email_check_match_email_addresses($message['from']);
+    // When no address is found, no need to try sending the email.
+    //
+    // If an address is found, just deal with the first one to determine if
+    // we need to set an Reply-To header.
+    if (count($message_from_addresses) === 0) {
+      watchdog('email_check', 'No mail address found in string "%from".', array('%from' => $message['from']), WATCHDOG_WARNING);
+      $message['send'] = FALSE;
+      return;
+    } else {
```
Comment (torotil) [inline email_check.module]: The trailing `{` should be on a separate line. (phpcs)

File: email_check.module [hunk 3]
```diff
@@ -21,47 +23,102 @@
...
+      // we know we need to set a reply-to. So save the found address as
+      // reply-to and use the site mail as From address (which we assume to be
+      // valid).
+      //
+      // Only set the address as reply-to if reply-to is not set already.
       if (strpos($address, '@' . $site_mail_domain) === FALSE) {
         $message_from_addresses[$index] = $site_mail_address;
         if (!$reply_to) {
           $reply_to = $address;
+        } else {
```
Comment (torotil) [inline email_check.module]: The `{` should be on a separate line.

File: email_check.module [hunk 4]
```diff
@@ -21,47 +23,102 @@
...
+    // If the replyto is set, use it.
+    //
+    // No need to check the domain of the reply to address, everything is
+    // allowed here.
     if ($reply_to) {
       $message['headers']['Reply-To'] = $reply_to;
     }
   }
+
+  // the Drupal default is to use the messages 'From' address
```
Comment (torotil) [inline email_check.module]: Comments should be whole sentences … (phpcs)

File: tests/MailAlterTest.php [hunk 5]
```diff
@@ -73,6 +73,175 @@ public function testNoReplacingWithExplicitReplyTo() {
...
+    $this->setConfig(['site_mail_return_path' => 'bounce@example.com']);
+    $message['from'] = 'some@example.com';
+    email_check_mail_alter($message);
+    $this->assertEqual('some@example.com', $message['from']);
+    $this->assertEqual('bounce@example.com', $message['headers']['Return-Path']);
+  }
+
+  /**
+   * Return-Path is only set when in mail domain.
+   * Default is the site_mail.
```
Comment (torotil) [inline tests/MailAlterTest.php]: Function doc summary should be 1 line (optionally) followed by a blank line and the rest of the doc. (phpcs)

There is many more instances of that in here.

File: tests/MailAlterTest.php [hunk 6]
```diff
@@ -73,6 +73,175 @@ public function testNoReplacingWithExplicitReplyTo() {
...
+   * The "primary address" i.e. the first determines the behaviour.
+   * The other from addresses are simply copied.
+   */
+  public function testMultipleHeaderFromAddresses() {
+    $message['from'] = 'some@example.com, another@example.com';
+    email_check_mail_alter($message);
+    $this->assertEqual('some@example.com,another@example.com', $message['from']);
+    $this->assertArrayNotHasKey('Reply-To', $message['headers']);
+  }
+  // treat as primary --> replyto is set
```
Comment (torotil) [inline tests/MailAlterTest.php]: Missing function doc. (phpcs) Many more instances of this in the tests.

**Review by torotil — CHANGES_REQUESTED** [review]
Thanks for fixing this and writing all those tests! Could you also fix the coding-style issues I’ve pointed out?

**Review by torotil — APPROVED** [review]

**Comment by torotil** [comment]
Never mind I’ve fixed the coding-style issues myself. @alex-mo Could you verify that the doc-strings I’ve modified are still correct?

**Comment by alex-mo** [comment]
Ah yeah. Forgot running `phpcs`...

**Comment by alex-mo** [comment]
Checked the comments and changed two: the `email_check` only looks for the first address (and not if any is in the mail domain), when multiple From-addresses are given...

**Pull request merged**